polar-core/src/polar.rs

```diff
@@ -211,35 +213,54 @@ impl Polar {
                     }
                 }
             }
-            // Rewrite shorthand rules in resource blocks before validating rule prototypes.
-            kb.rewrite_shorthand_rules()?;
-            // check rules are valid against rule prototypes
-            kb.validate_rules()?;
             Ok(warnings)
         }
 
-        // if any of the lines fail to load, we need to remove the source from
-        // the knowledge base
-        match load_source(source_id, &source, &mut kb) {
-            Ok(warnings) => {
-                self.messages.extend(warnings.iter().map(|m| Message {
-                    kind: MessageKind::Warning,
-                    msg: m.to_owned(),
-                }));
-                Ok(())
-            }
-            Err(e) => {
-                kb.remove_source(source_id);
-                Err(e)
+        let mut kb = self.kb.write().unwrap();
+        for source in &sources {
+            let result = kb.add_source(source.clone());
+            let result = result.and_then(|source_id| load_source(source_id, source, &mut kb));
+            match result {
+                Ok(warnings) => {
+                    let warnings = warnings.into_iter().map(|msg| Message {
+                        kind: MessageKind::Warning,
+                        msg,
+                    });
+                    self.messages.extend(warnings);
+                }
+                Err(e) => {
+                    // If any source fails to load, clear the KB.
+                    kb.clear_rules();
+                    return Err(e);
+                }
             }
         }
+
+        // Rewrite shorthand rules in resource blocks before validating rule prototypes.
+        if let Err(e) = kb.rewrite_shorthand_rules() {
+            // If rewriting shorthand rules fails, clear the KB.
+            kb.clear_rules();
+            return Err(e);
+        }
+
+        // check rules are valid against rule prototypes
+        if let Err(e) = kb.validate_rules() {
+            // If rule type validation fails, clear the KB.
+            kb.clear_rules();
+            return Err(e);
+        }
+        Ok(())
     }
 
     // Used in integration tests
     pub fn load_str(&self, src: &str) -> PolarResult<()> {
-        self.load(src, None)
+        self.load(vec![Source {
+            src: src.to_owned(),
+            filename: None,
+        }])
     }
 
+    // TODO(gj): ask Sam if we still need this.
```

Comment:
@samscott89

Reply:
We might need to rethink the strategy in polar-analyzer after these changes. We may as well keep it for now though.

Reply:
Okay, works for me.